core/src/common.cpp

```diff
@@ -362,6 +364,439 @@ namespace irods::http
 		return std::nullopt;
 	}
 
+	/// Validates an OAuth 2.0 Access Token using the Introspection Endpoint
+	/// See RFC 7662 for more details
+	///
+	/// \returns An optional containing a nlohmann::json object if verification was successful. Otherwise,
+	///          an empty optional is returned.
+	auto validate_using_introspection_endpoint(std::string _bearer_token) -> std::optional<nlohmann::json>
+	{
+		namespace logging = irods::http::log;
+
+		body_arguments args{{"token", _bearer_token}, {"token_type_hint", "access_token"}};
+
+		auto json_res{hit_introspection_endpoint(url_encode_body(args))};
+
+		// Validate access token
+		if (!json_res.at("active").get<bool>()) {
+			logging::warn("{}: Access token is invalid or expired.", __func__);
+			return std::nullopt;
+		}
+
+		return json_res;
+	}
+
+	/// Fetches JWKs from the location specified by the OpenID Provider
+	///
+	/// See OpenID Connect Discovery 1.0 Section 3 for info on jwks_uri
+	/// See RFC 7517 for more information on JSON Web Key (JWK)
+	///
+	/// \returns A std::string represting the JWKs from the OpenID Provider
+	///
+	/// \todo Stash the results from this function, only allow to run once.
+	auto fetch_jwks_from_openid_provider() -> std::string
+	{
+		namespace logging = irods::http::log;
+
+		const auto jwks_uri{
+			irods::http::globals::oidc_endpoint_configuration().at("jwks_uri").get_ref<const std::string&>()};
+
+		const auto parsed_uri{boost::urls::parse_uri(jwks_uri)};
+
+		if (parsed_uri.has_error()) {
+			logging::error("{}: Error trying to parse jwks_uri [{}]. Please check configuration.", __func__, jwks_uri);
+			return "{\"error\", \"bad endpoint\"}";
+		}
```

Comment:
Consider throwing here instead.

If we return, the error will be stashed in a static variable...

Reply:
Done.